### PR TITLE
Hotfixes for the auth-service (forgot password feature)

### PR DESCRIPTION
**Pull request by Baalmart**
**_WHAT DOES THIS PR DO?_**

- [x] Align some identified error response with the organisational format
- [x] Fix an error with the forgot password feature. Focus on the `update password via email` endpoint
- [x] For identified cases, ensure that we always return the right http status codes in all use cases. reference Belinda Slack messages for 20th March 2023.
- [x] looks like the  input validation defaults to bad gateway when one enters `emails` instead of `email`, the input validation needs to be addressed.
- [x] Include datetime validity to the verify access token endpoint

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
    - [AN-323]
- Issues
    - https://github.com/airqo-platform/AirQo-api/issues/1411

**_HOW DO I TEST OUT THIS PR?_**
```
cd src/auth-service
npm install
npm run dev-mac
```

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [x] [forgot password](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-403982cf-4bb1-4b6a-922a-9fb13b683e89)
- [x] [update password via email](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-5f2cdfe3-7381-47e9-ab6b-65bf7306e0d2)

[AN-323]: https://airqoteam.atlassian.net/browse/AN-323?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/auth-service/models/User.js [hunk 1]
```diff
@@ -325,19 +325,18 @@ UserSchema.statics = {
...
-          data,
+          data: response,
           status: httpStatus.OK,
         };
       } else if (isEmpty(response)) {
         return {
           success: true,
           message: "no users exist",
           data: [],
-          status: httpStatus.NOT_FOUND,
```
Comment (Baalmart) [inline src/auth-service/models/User.js:340]: correcting the http status code or cases of NO records

File: src/auth-service/utils/create-user.js [hunk 2]
```diff
@@ -468,40 +455,13 @@ const join = {
...
-              status: responseFromSendEmail.status
-                ? responseFromSendEmail.status
-                : "",
-            };
+            return responseFromSendEmail;
           }
         } else if (responseFromSaveToken.success === false) {
-          return {
-            success: false,
-            message: responseFromSaveToken.message,
```
Comment (Baalmart) [inline src/auth-service/utils/create-user.js:485]: just simple refactoring to reduce on lines of code and also align all the error responses to the AirQo standard of JSON response =`{errors: {message:"error message"}}`

File: src/auth-service/utils/create-user.js [hunk 3]
```diff
@@ -762,46 +673,16 @@ const join = {
           "responseFromUpdateUser in update forgotten password",
           responseFromUpdateUser
         );
-        if (responseFromUpdateUser.success === true) {
-          return {
-            success: true,
-            message: responseFromUpdateUser.message,
```
Comment (Baalmart) [inline src/auth-service/utils/create-user.js:768]: just simple refactoring to reduce on lines of code and also align all the error responses to the AirQo standard of JSON response =`{errors: {message:"error message"}}`

File: src/auth-service/utils/generate-filter.js [hunk 4]
```diff
@@ -29,9 +29,9 @@ const filter = {
       }
       if (resetPasswordToken) {
         filter["resetPasswordToken"] = resetPasswordToken;
-        filter["resetPasswordExpires"] = {
-          $gt: Date.now(),
-        };
+        // filter["resetPasswordExpires"] = {
```
Comment (Baalmart) [inline src/auth-service/utils/generate-filter.js:32]: fixing the forgot password bug from here. @BenjaminSsempala @Codebmk 

Apparently, the system is not able to do datetime comparisons when querying from the mongo DB.

File: src/auth-service/utils/request-access.js [hunk 5]
```diff
@@ -403,32 +382,12 @@ const requestAccess = {
...
   delete: async (tenant, filter) => {
     try {
-      let responseFromRemoveCandidate = await CandidateModel(
+      const responseFromRemoveCandidate = await CandidateModel(
         tenant.toLowerCase()
       ).remove({
         filter,
       });
-
-      if (responseFromRemoveCandidate.success == true) {
```
Comment (Baalmart) [inline src/auth-service/utils/request-access.js:412]: just simple refactoring to reduce on lines of code and also align all the error responses to the AirQo standard of JSON response =`{errors: {message:"error message"}}`

**Review by 123MwanjeMike — APPROVED** [review]
Thanks @Baalmart 
LGTM

image

File: src/auth-service/utils/create-user.js [hunk 6]
```diff
@@ -682,126 +634,74 @@ const join = {
...
-        tenant.toLowerCase(),
-        filter
-      );
+      const { tenant, body } = request;
+      const { resetPasswordToken } = body;
+      const timeZone = moment.tz.guess();
+      let filter = {
+        resetPasswordToken,
+        resetPasswordExpires: {
+          $gt: moment().tz(timeZone).toDate(),
```
Comment (Baalmart) [inline src/auth-service/utils/create-user.js:668]: this is the most recent addition to address timezone differences of the server which was the causing the datetime failure when updating forgotten passwords. Perhaps the recent changes in server settings led to this challenge which has been fixed with this module. @BenjaminSsempala @123MwanjeMike @Codebmk @NoahNsimbe

File: src/auth-service/utils/control-access.js [hunk 7]
```diff
@@ -121,10 +122,13 @@ const controlAccess = {
...
       const limit = parseInt(request.query.limit, 0);
       const skip = parseInt(request.query.skip, 0);
-
+      const timeZone = moment.tz.guess();
       let filter = {
         token,
         user_id,
+        expires: {
+          $gt: moment().tz(timeZone).toDate(),
+        },
```
Comment (Baalmart) [inline src/auth-service/utils/control-access.js:131]: @Codebmk / @123MwanjeMike 
checks for the validity of the `access token`. This default validity is added in the access token creation via `config.constants`:

```
1.  EMAIL_VERIFICATION_HOURS: 1,
2.  EMAIL_VERIFICATION_MIN: 0,
3.  EMAIL_VERIFICATION_SEC: 0,
```

File: src/auth-service/controllers/create-user.js [hunk 8]
```diff
@@ -155,12 +156,15 @@ const createUser = {
       const { email, phoneNumber, uid, providerId, providerUid } = req.body;
       const hasErrors = !validationResult(req).isEmpty();
       if (hasErrors) {
+        logObject("hasErrors", hasErrors);
         let nestedErrors = validationResult(req).errors[0].nestedErrors;
+
         logger.error(
           `input validation errors ${JSON.stringify(
-            manipulateArraysUtil.convertErrorArrayToObject(nestedErrors)
```
Comment (Baalmart) [inline src/auth-service/controllers/create-user.js:161]: the runtime issue which was causing the gateway timeouts (with bad input) whenever user was looking up users from Firebase

**Pull request merged**